// image-converter@1.0.0/karma.conf.js
module.exports = function(config) {
  config.set({

    basePath: '',

    frameworks: ['jasmine'],
    plugins:[
      'karma-webpack',
      'karma-jasmine',
      'karma-chrome-launcher'
    ],

    files: [
      'spec/svg.spec.js'
    ],

    exclude: [
    ],

    preprocessors: {
      'spec/*.js':['webpack'],
      'app/*.js':['webpack']
    },
    webpack:{
      entry:'./spec/svg.spec.js',
      output : {
        filename: 'bundle.js'
      }
    },

    reporters: ['progress'],

    port: 9876,



    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.DEBUG,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],
    

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity
  })
}
